libsolidity/analysis/TypeChecker.cpp

```diff
@@ -426,6 +429,104 @@ bool TypeChecker::visit(FunctionDefinition const& _function)
+		{
+			auto&& mantissaType = dynamic_cast<IntegerType const*>(_function.parameterList().parameters()[0]->type());
+			auto&& exponentType = dynamic_cast<IntegerType const*>(_function.parameterList().parameters()[1]->type());
+
+			vector<string> mantissaOrExponentTypeErrorMessages;
+			if (!mantissaType)
+				mantissaOrExponentTypeErrorMessages.emplace_back("The mantissa parameter must be an integer.");
+			if (!exponentType)
+				mantissaOrExponentTypeErrorMessages.emplace_back("The exponent parameter must be an unsigned integer.");
+
+			if (!mantissaOrExponentTypeErrorMessages.empty())
+				m_errorReporter.typeError(
+					1587_error,
+					_function.parameterList().location(),
+					"Literal suffix function has invalid parameter types. " +
+					joinHumanReadable(mantissaOrExponentTypeErrorMessages, " ")
+				);
+
+			if (exponentType && exponentType->isSigned())
+				m_errorReporter.typeError(
+					3123_error,
+					_function.parameterList().parameters()[1]->typeName().location(),
+					"The exponent parameter of a literal suffix function must be unsigned. "
+					"Exponent is always either zero or a negative power of 10 but the parameter represents its absolute value."
+				);
+		}
+		else if (_function.parameterList().parameters().size() == 1)
+		{
+			auto&& parameterType = _function.parameterList().parameters()[0]->type();
+
+			if (dynamic_cast<FixedPointType const*>(parameterType))
+				m_errorReporter.typeError(
+					2699_error,
+					_function.parameterList().parameters()[0]->location(),
+					"Parameters of fixed-point types are not allowed in literal suffix functions. "
+					"To support fractional literals the suffix function must accept two integer arguments "
+					"(mantissa and exponent) that such literals can be decomposed into."
+				);
+
+			if (
+				!TypeProvider::boolean()->isImplicitlyConvertibleTo(*parameterType) &&
+				// ASSUMPTION: There are no address payable literals.
+				!TypeProvider::address()->isImplicitlyConvertibleTo(*parameterType) &&
+				// ASSUMPTION: Literal 1 is implicitly convertible to any integer type.
+				!TypeProvider::rationalNumber(1)->isImplicitlyConvertibleTo(*parameterType) &&
+				// ASSUMPTION: bytes1 is implicitly convertible to any fixed-bytes type.
+				!TypeProvider::fixedBytes(1)->isImplicitlyConvertibleTo(*parameterType) &&
+				!TypeProvider::stringLiteral("a")->isImplicitlyConvertibleTo(*parameterType)
+			)
+				m_errorReporter.typeError(
+					2998_error,
+					_function.parameterList().parameters()[0]->location(),
+					"This literal suffix function is not usable as a suffix because no literal is "
+					"implicitly convertible to its parameter type."
+				);
+		}
+
+		solAssert(_function.returnParameterList());
+		if (_function.returnParameterList()->parameters().size() != 1)
+		{
+			m_errorReporter.typeError(
+				7848_error,
+				_function.returnParameterList()->location(),
+				"Literal suffix functions must return exactly one value."
+			);
+		}
```

Comment:
```suggestion
			m_errorReporter.typeError(
				7848_error,
				_function.returnParameterList()->location(),
				"Literal suffix functions must return exactly one value."
			);
```
If I'm allowed to be nitpicky as well this time ;-).

Reply:
Nitpick as much as you want. This kind of stuff is trivial to fix while I'm fixing other stuff anyway.